File: system-test/java/src/main/java/com/optivem/eshop/dsl/driver/adapter/shop/api/client/controllers/OrderController.java
package com.optivem.eshop.dsl.driver.adapter.shop.api.client.controllers;

import com.optivem.eshop.dsl.driver.port.shop.dtos.ViewOrderResponse;
import com.optivem.eshop.dsl.driver.port.shop.dtos.PlaceOrderRequest;
import com.optivem.eshop.dsl.driver.port.shop.dtos.PlaceOrderResponse;
import com.optivem.eshop.dsl.driver.adapter.shop.api.client.dtos.errors.ProblemDetailResponse;
import com.optivem.eshop.dsl.driver.adapter.shared.client.http.JsonHttpClient;
import com.optivem.eshop.dsl.common.Result;

public class OrderController {
    private static final String ENDPOINT = "/api/orders";

    private final JsonHttpClient<ProblemDetailResponse> httpClient;

    public OrderController(JsonHttpClient<ProblemDetailResponse> httpClient) {
        this.httpClient = httpClient;
    }

    public Result<PlaceOrderResponse, ProblemDetailResponse> placeOrder(PlaceOrderRequest request) {
        return httpClient.post(ENDPOINT, request, PlaceOrderResponse.class);
    }

    public Result<ViewOrderResponse, ProblemDetailResponse> viewOrder(String orderNumber) {
        return httpClient.get(ENDPOINT + "/" + orderNumber, ViewOrderResponse.class);
    }

}

File: system-test/typescript/src/testkit/driver/adapter/api/client/controllers/OrderController.ts
import type { Result } from '../../../../../common/result.js';
import type { PlaceOrderRequest } from '../../../../port/dtos/PlaceOrderRequest.js';
import type { PlaceOrderResponse } from '../../../../port/dtos/PlaceOrderResponse.js';
import type { ViewOrderResponse } from '../../../../port/dtos/ViewOrderResponse.js';
import type { SystemError } from '../../../../port/dtos/errors/SystemError.js';
import { JsonHttpClient } from '../../../shared/client/http/json-http-client.js';
import { SystemErrorMapper } from '../../SystemErrorMapper.js';

export class OrderController {
  private static readonly ENDPOINT = '/api/orders';

  private readonly httpClient: JsonHttpClient<SystemError>;

  constructor(baseUrl: string) {
    this.httpClient = new JsonHttpClient(baseUrl, SystemErrorMapper.fromResponse);
  }

  placeOrder(request: PlaceOrderRequest): Promise<Result<PlaceOrderResponse, SystemError>> {
    return this.httpClient.post<PlaceOrderResponse>(OrderController.ENDPOINT, request);
  }

  viewOrder(orderNumber: string): Promise<Result<ViewOrderResponse, SystemError>> {
    return this.httpClient.get<ViewOrderResponse>(`${OrderController.ENDPOINT}/${orderNumber}`);
  }

  cancelOrder(orderNumber: string): Promise<Result<void, SystemError>> {
    return this.httpClient.postVoid(`${OrderController.ENDPOINT}/${orderNumber}/cancel`, {});
  }

  deliverOrder(orderNumber: string): Promise<Result<void, SystemError>> {
    return this.httpClient.postVoid(`${OrderController.ENDPOINT}/${orderNumber}/deliver`, {});
  }
}
